Add refined label box

robobot3000 opened the pull request:
Adds #563 

Please test with the latest `master` server checked out.

* Visual changes
* More info added to the label boxes, including the refund status, label service, package and its contents
* Added a history meta box logging purchases, refunds and reprints

CC @allendav @kellychoffman @DanReyLop @nabsul @jeffstieler

robobot3000 commented:
Screenshot:
<img width="379" alt="screen shot 2016-11-29 at 14 22 30" src="https://cloud.githubusercontent.com/assets/800604/20709670/5b3514be-b63f-11e6-83c6-2f3dec4c764f.png">

Test tips:
* purchase a bunch of labels
* request a refund
  * _some refunds return an error saying that the parcel has been shipped or the refund has already been requested. @jeffstieler - is this something to be worried about? This PR doesn't change the body of the refund request, and I'm assuming this is because the label IDs are reused in the test environment_
* in the server DB, modify the rows in `wcc_shipping_label_refund` table for your pending refunds. Set the `status` value to 1 for pending, 3 for rejected, 6 for complete (+ set the `refund_date` value)
* refresh the page and you should see the status
  * note that only the status for labels that are marked as pending is refreshed. If the refund has been marked as complete or rejected once, it won't be updated from the server
* confirm that everything works as expected

jeffstieler commented:
> some refunds return an error saying that the parcel has been shipped or the refund has already been requested. @jeffstieler - is this something to be worried about? This PR doesn't change the body of the refund request, and I'm assuming this is because the label IDs are reused in the test environment

@robobot3000 the "parcel has been shipped" will occur randomly, and is expected. EasyPost just updates any refunds in the test enviroment with random statuses.

You should *not* be seeing "refund already requested" messages for new labels. That is based on the label ID being found in the refunds table. Can you reliably reproduce this?

robobot3000 commented:
> You should not be seeing "refund already requested" messages for new labels. That is based on the label ID being found in the refunds table. Can you reliably reproduce this?

It happens only when I request a refund right after getting the "parcel has been shipped" message. I assume it's to be expected as well

jeffstieler commented:
> It happens only when I request a refund right after getting the "parcel has been shipped" message. I assume it's to be expected as well

Yep - that's expected if you're requesting a refund again for the same label.

jeffstieler commented:
Just in testing (haven't gotten into code yet):

1. I tried to refund a label, and got the error that the package had already been shipped. The label history still showed that I refunded the label.

Can we check the response of the refund request before adding to history? It might still make sense to add an item to the history, but it should indicate that the refund was *not* successful.

2. When I take an action that gets added to the history, the immediate UI update just says "User" instead of my username:

![screen shot 2016-12-01 at 10 36 59 am](https://cloud.githubusercontent.com/assets/63922/20805215/2572a9de-b7b3-11e6-84f3-eddbfc737a9d.png)

robobot3000 commented:
0. I added label IDs to the UI, so it's easier to read the history modal and see which labels had actions performed on them
1. I changed the history entry to appear as `<user> unsuccessfully requested a refund of Label #95` when the request errors out, and similar when the request was successful
2. The reprint history entry should now display the name/url properly without the need for page refresh

jeffstieler commented on classes/class-wc-connect-service-settings-store.php:
```diff
@@ -122,6 +122,71 @@ public function update_label_order_meta_data( $order_id, $new_label_data ) {
+				$history_data = json_decode( $raw_history_data, true, WOOCOMMERCE_CONNECT_MAX_JSON_DECODE_DEPTH );
+			} else {
+				$history_data = array();
+			}
```
This looks like a candidate for DRY. (it's repeated from above)

jeffstieler commented on classes/class-wc-connect-shipping-label.php:
```diff
@@ -301,27 +301,34 @@ public function meta_box( $post ) {
 				'addressNormalizationURL' => get_rest_url( null, '/wc/v1/connect/normalize-address' ),
 				'getRatesURL'             => get_rest_url( null, '/wc/v1/connect/shipping-rates' ),
-				'labelStatusURL'          => get_rest_url( null, '/wc/v1/connect/label/' . $order->id . '-%d' ),
+				'labelStatusURL'          => get_rest_url( null, '/wc/v1/connect/label/' . $order->id . '-%d-%d' ),
```
It looks like you're following the existing pattern, but why are these endpoint parameters part of the path? Why aren't we using a query string here, or posting some JSON?

jeffstieler commented on classes/class-wc-rest-connect-shipping-label-controller.php:
```diff
@@ -101,21 +109,48 @@ public function update_items( $request ) {
+
+			$product_names = array();
+			foreach ( $package[ 'products' ] as $product_id ) {
+				$product = $product_factory->get_product( $product_id );
```
You should be able to use [`wc_get_product()`](https://docs.woocommerce.com/wc-apidocs/function-wc_get_product.html) here.

jeffstieler commented on classes/class-wc-rest-connect-shipping-label-controller.php:
```diff
@@ -89,8 +91,14 @@ public function update_items( $request ) {
+			$labels_order_meta = json_decode( $existing_labels_data, true, WOOCOMMERCE_CONNECT_MAX_JSON_DECODE_DEPTH );
+		}
+		$package_lookup = $this->settings_store->get_package_lookup_for_service( $carrier );
+		$product_factory = new WC_Product_Factory();
```
This should be unnecessary, see comment below.

jeffstieler commented on classes/class-wc-rest-connect-shipping-label-controller.php:
```diff
@@ -101,21 +109,48 @@ public function update_items( $request ) {
+			} else if ( isset( $package_lookup[ $box_id ] ) ) {
+				$label_meta[ 'package_name' ] = $package_lookup[ $box_id ][ 'name' ];
+			} else {
+				$label_meta[ 'package_name' ] = __( 'Unknown package', 'connectforwoocommerce' );
```
Can you explain the circumstances that lead to an "unknown" package?

robobot3000 replied:
It's kind of an edge case, but it's possible to print a label for a box that has been removed from a packaging manager.

jeffstieler commented on classes/class-wc-connect-service-settings-store.php:
```diff
@@ -122,6 +122,71 @@ public function update_label_order_meta_data( $order_id, $new_label_data ) {
+			return $entry;
+		}
+
+		public function update_label_order_history( $order_id, $label_ids, $entry_type, $time ) {
```
All uses of this pass in `$time = time() * 1000`. Perhaps this can just be an optional argument with a default?

Additionally, what do you think about handling both single `$label_id` and an array of `$label_ids`? That way, function calls don't have to wrap `array( $a_single_label_id )`.

jeffstieler commented on classes/class-wc-rest-connect-shipping-label-status-controller.php:
```diff
@@ -22,7 +22,7 @@ class WC_REST_Connect_Shipping_Label_Status_Controller extends WP_REST_Controlle
 	 * @var string
 	 */
-	protected $rest_base = 'connect/label/(?P<order_id>\d+)-(?P<label_id>\d+)';
+	protected $rest_base = 'connect/label/(?P<order_id>\d+)-(?P<label_id>\d+)-(?P<get_refund>\d+)';
```
Same question about path vs. query string.

jeffstieler commented on client/shipping-label/views/history.js:
```diff
@@ -0,0 +1,79 @@
+	};
+
+	const renderHistoryEntry = ( entry, index ) => {
+		const entryMoment = CalypsoI18n.moment( entry.time ).fromNow();
```
Are we using Calypso's `moment()` anywhere else? I'm not sure we should be here.. from what I understand, `i18n-calypso` exposes a `moment` instance with a locale already set, but if we aren't using `i18n-calypso` for all translations in the project - can't this be a mismatch?

Why not use `moment` directly?

jeffstieler commented on client/shipping-label/views/history.js:
```diff
@@ -0,0 +1,79 @@
+		document.getElementById( 'woocommerce-order-label-history' ).style.display = 'none';
+		return null;
+	}
+	document.getElementById( 'woocommerce-order-label-history' ).style.display = 'block';
```
_This comment applies to both uses of `style.display =` above_

Shouldn't this be applied with a class that gets dynamically set based on the value of `labelHistory`?

robobot3000 replied:
This has to hide the whole meta box, which is not controlled by React, but by PHP. I could use a class, but it is easier and more cross-browser compatible to just set the CSS `display` directly.

jeffstieler replied:
I think we should be hiding the history metabox by default (in our stylesheet) and overriding the `display` property if there's history to be shown. I'm willing to concede the class-based approach, but are you *sure* that using a class to set `display` isn't cross browser compatible?

See: https://developer.mozilla.org/en-US/docs/Web/CSS/display#Browser_compatibility

I'd also like to see this logic moved to where the `LabelHistory` component is actually hooked into the DOM. (in the `ShippingLabelRootView` component)

(cc: @allendav @DanReyLop @v18 @nabsul for discussion)

robobot3000 replied:
The history metabox is not empty - it contains markup generated by `add_meta_box`. But I added CSS to hide it by default and some code to show it.

By browser compatibility I meant JavaScript's `classList.add()`, but the only browser that doesn't support it is IE7, so it turns out everything is good.

jeffstieler commented on client/shipping-label/views/index.js:
```diff
@@ -1,112 +1,222 @@
+	}
+
+	renderLabel( label, index ) {
+		const purchased = CalypsoI18n.moment( label.created ).fromNow();
```
Same question about using `i18n-calypso` here.

jeffstieler commented on client/shipping-label/views/index.js:
```diff
@@ -1,112 +1,222 @@
+	}
+
+	componentDidMount() {
+		this.historyMetaBox = document.getElementById( 'wc-connect-shipping-label-history' );
```
Is there a reason this needs to be a separate element from the label purchase metabox?

(This is the first occurrence of `document.getElementById()` and `ReactDOM.render()` outside `main.js`, so it stood out to me)

robobot3000 replied:
The only reason is the design - it shows the history in a separate metabox. I think visually it's cleaner, but unfortunately, this is what needs to be done in order to achieve it in the code.

DanReyLop replied:
The other meta-box should be another completely different React "application", but currently we rely on a global variable `wcConnectData` so it's not possible. Tomorrow I have a ✈️ , I'll give this a shot and see if I can make 2 applications live in the same page.

jeffstieler commented on client/shipping-label/views/index.js:
```diff
@@ -1,112 +1,222 @@
+		const thirtyDaysAgo = new Date().setDate( today.getDate() - 30 );
+		if ( ( label.used_date && label.used_date < today.getTime() ) || ( label.created_date && label.created_date < thirtyDaysAgo ) ) {
+			return null;
+		}
```
Are we sure we want to enforce "refundability" logic in this component? Does it make more sense on the server perhaps?

If we do - I think a label having a `used_date` at all makes it ineligible for refund.

robobot3000 replied:
You're right. It makes sense for this to live in the Connect server. I opened Automattic/woocommerce-connect-server#711 and I think it could be handled in a separate PR.

jeffstieler commented on client/shipping-label/views/index.js:
```diff
@@ -1,112 +1,222 @@
+				text = sprintf( __( 'Refunded on %s' ), formatDate( label.refund.refund_date ) );
+				break;
+			case 'rejected':
+				className = ( 'wcc-metabox-label-item__refund-rejected' );
```
Nitpick: why the surrounding parentheses?

jeffstieler commented on client/shipping-label/views/index.js:
```diff
@@ -1,112 +1,222 @@
+		const todayTime = new Date().getTime();
+		if ( ( label.used_date && label.used_date < todayTime ) || ( label.expiry_date && label.expiry_date < todayTime ) ) {
+			return null;
+		}
```
It makes sense to inquire about where this logic should live, given my comments about refund links..

robobot3000 replied:
See above: Automattic/woocommerce-connect-server#711

jeffstieler commented on woocommerce-connect-client.php:
```diff
@@ -747,9 +748,15 @@ public function hide_wc_connect_package_meta_data( $hidden_keys ) {
+					break;
+				case 'wc_connect_label_history':
+					$protected = true;
+					break;
```
You can stack `switch` `case`s that have the same resulting block:
```javascript
switch ( $meta_key ) {
	case 'wc_connect_labels':
	case 'wc_connect_label_history':
		$protected = true;
		break;
}
```

Basically, if you remove the `break;`, the execution path just "falls through" to the next case.

jeffstieler commented:
Sorry for the delay on the review - this is a big PR!

allendav commented:
Needs some styling tweaks - the fonts are a bit bigger than the design, spacing too

cc @kellychoffman 

i.e. (for each pairing, the top is the design, the bottom this PR)

<img width="544" alt="screen shot 2016-12-07 at 1 17 49 pm" src="https://cloud.githubusercontent.com/assets/1595739/20986990/bba8ac7a-bc7f-11e6-9918-3a499dd79fc3.png">

allendav commented:
Let's suppress outline on the tracking number link, i.e. get rid of this effect after clicking:

<img width="297" alt="screen shot 2016-12-07 at 1 27 19 pm" src="https://cloud.githubusercontent.com/assets/1595739/20987390/49f108a0-bc81-11e6-970a-c37f4a9510dd.png">

robobot3000 commented:
Thanks for having a look @jeffstieler @allendav 

I pushed some code and style changes.

jeffstieler commented on client/lib/utils/time-ago.js:
```diff
@@ -0,0 +1,48 @@
+import { translate as __ } from 'lib/mixins/i18n';
+import { sprintf } from 'sprintf-js';
+
```
I should have been more clear - I wasn't against the use of `moment()`, I just questioned bringing in all of Calypso's `i18n` module to use it.

Don't feel like you have to remove this now though.

robobot3000 replied:
I decided to create my own - from what I read it is difficult to get `moment` to use the correct locale when it's installed with npm, and it seemed that we are not using large parts of that library.

DanReyLop replied:
I suggest you take a look at `date-fns`: https://date-fns.org/docs/distanceInWordsToNow
I18N in `date-fns`: https://date-fns.org/docs/I18n

It's being advertised as "similar to `moment`, but you can just use individual functions". It certainly looks like a better solution than reimplementing it yourself :)

kellychoffman commented:
Why are there two sections of label history? 

<img width="298" alt="screen shot 2016-12-08 at 12 20 12 pm" src="https://cloud.githubusercontent.com/assets/5835847/21025033/24c38e18-bd44-11e6-9767-2fbc4b6a960a.png">

1. We should get rid of the second meta box below, as it is redundant and not as informative as the listing above.
2. Add a 'Shipping labels history' header above the history listing that only displays when there is history?

kellychoffman commented:
When a refund is pending on a label, you shouldn't be able to reprint the label. Let's remove that link. 

<img width="296" alt="screen shot 2016-12-08 at 12 50 59 pm" src="https://cloud.githubusercontent.com/assets/5835847/21025282/1106c2cc-bd45-11e6-986d-a07a07ac1d26.png">

robobot3000 commented:
I hid the reprint button for the labels with refund requested.

Re: two meta boxes:

This is what has been requested by @allendav in #562:
> a reverse chronologically sorted action log of create, reprint, and refund actions with date and username of action initiator

And what has been added to the design by @jameskoster:
<img src="https://camo.githubusercontent.com/f1288f549370d6c3f9573b5e87491d96c968af3c/68747470733a2f2f636c6475702e636f6d2f58546c4e6e395f7753342e706e67" />

Are you sure you want me to remove it?

kellychoffman commented:
Ah, I was going off of the latest design in that thread: https://github.com/Automattic/woocommerce-connect-client/issues/562#issuecomment-256627258

I think the double history listing is too confusing.

allendav commented:
@robobot3000 - I see the disconnect about the second meta box - i did not even notice it yesterday because it appears way below the fold on my Edit Order pages.

As I re-read my description in the original issue I can see how the reverse chronology could be interpreted as a distinct, separate component but I didn't mean it that way - my apologies for the misunderstanding

Let's go ahead and remove the 2nd metabox and keep just the richer top container

allendav commented:
Thank you @kellychoffman for the style updates

@robobot3000 - delete the Shipping Labels History meta box and then this is good to go

kellychoffman commented:
👍 Looks good.
